tests/federation_room_join_partial_state_test.go

```diff
@@ -489,6 +490,51 @@ func testReceiveEventDuringPartialStateJoin(
 		t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
 	}
 
+	// fire off a /state_ids request for the last event.
+	// it must either:
+	//   * block because the homeserver does not have full state at the last event
+	//   * or 403 because the homeserver does not have full state yet and does not consider the
+	//     Complement homeserver to be in the room
+	stateIdsResponseChan := make(chan *gomatrixserverlib.RespStateIDs)
+	defer close(stateIdsResponseChan)
+	go func() {
+		stateReq := gomatrixserverlib.NewFederationRequest("GET", "hs1",
+			fmt.Sprintf("/_matrix/federation/v1/state_ids/%s?event_id=%s",
+				url.PathEscape(psjResult.ServerRoom.RoomID),
+				url.QueryEscape(event.EventID()),
+			),
+		)
+		var respStateIDs gomatrixserverlib.RespStateIDs
+		if err := psjResult.Server.SendFederationRequest(deployment, stateReq, &respStateIDs); err != nil {
+			httpErr, ok := err.(gomatrix.HTTPError)
+			t.Logf("%v", httpErr)
+			if ok && httpErr.Code == 403 {
+				stateIdsResponseChan <- nil
+				return
+			}
+			t.Errorf("/state_ids request returned non-200: %s", err)
+			return
+		}
+		stateIdsResponseChan <- &respStateIDs
```

Comment:
rather than matching for the 403 here and sending a `nil` magic value down the channel, I suggest you define 

```go
type stateIDsResult struct {
    RespStateIDs gomatrixserverlib.RespStateIDs
    Error error
}
```

... and then send that back down the channel, and do the matching in the main thread.